File: Aman_Bhaiya_DSA_List/Arrays/LQ22.java
import java.util.ArrayList;

public class LQ22 {
    public int trap(int[] height) {
        int n = height.length;
        ArrayList<Integer> left = new ArrayList<>();
        ArrayList<Integer> right = new ArrayList<>();
        int mn, lmx = height[0], rmx = height[n-1], total = 0;
        
        left.add(lmx);
        for(int i=1; i < n; i++){
            lmx = Math.max(lmx,height[i]);
            left.add(lmx);
        }

        right.add(rmx);
        for(int i = (n-2); i>=0; i--){
            rmx = Math.max(rmx,height[i]);
            right.add(rmx);
        }

        for(int i = 0; i < n; i++){
            mn = Math.min(left.get(i),right.get(n-i-1));
            total = total + (mn - height[i]);
        }

        return total;
    }
}

File: Aman_Bhaiya_DSA_List/Arrays/LQ22.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        total = 0
        left = []
        right = []
        lmx = height[0]
        rmx = height[n-1]

        left.append(lmx)
        for i in height[1:]:
            lmx = max(lmx,i)
            left.append(lmx)

        right.append(rmx)
        for i in height[n-2::-1]:
            rmx = max(rmx,i)
            right.append(rmx)
        
        for i in range(0,n):
            mn = min(left[i], right[n-i-1])
            total = total + (mn - height[i])

        return total